File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/components/meta/BotComponents.java
package org.firstinspires.ftc.teamcode.components.meta;

/* ?
? * Bot Component
? * Big bot component where all components all assembled
 * TODO: N/A
? */


import com.qualcomm.robotcore.hardware.HardwareMap;

import org.firstinspires.ftc.robotcore.external.Telemetry;
import org.firstinspires.ftc.teamcode.components.hardware.*;

public abstract class BotComponents {

    // Components for both TeleOp and Auton
    protected final Arm arm;
    protected final Claw claw;
    protected final Lift lift;

    protected final Wrist wrist;


    public BotComponents(HardwareMap hardwareMap, Telemetry telemetry) {
        this.arm = new Arm(hardwareMap, telemetry);
        this.lift = new Lift(hardwareMap, telemetry);
        this.claw = new Claw(hardwareMap, telemetry);
        this.wrist = new Wrist(hardwareMap, telemetry);

    }

    // Abstract method to be implemented by subclasses
    public abstract void updateComponents(boolean useLiftDeadzone);
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/components/meta/botComponents.kt
package org.firstinspires.ftc.teamcode.components.meta

import com.qualcomm.hardware.rev.RevColorSensorV3
import ftc.rogue.blacksmith.BlackOp.Companion.hwMap
import ftc.rogue.blacksmith.util.kt.invoke
//import org.firstinspires.ftc.teamcode.roadrunner.drive.SampleMecanumDrive
import org.firstinspires.ftc.teamcode.components.*
import org.firstinspires.ftc.teamcode.components.hardware.drivetrain


abstract class BaseBotComponents {
//    val claw   = Claw()
//    val intake = Intake()
//    val arm    = Arm()
//    val wrist  = Wrist()
//    val lift   = Lift()

    open fun updateComponents(useLiftDeadzone: Boolean) {
//        claw.update()
//        arm.update()
//        wrist.update()
    }
}

fun createTeleOpBotComponents() =
    TeleOpBotComponents(
        hwMap(deviceNames.COLOR_SENSOR),
        drivetrain(),
    )

data class TeleOpBotComponents(
    val rcs: RevColorSensorV3,
    val drivetrain: drivetrain,
) : BaseBotComponents() {
    override fun updateComponents(useLiftDeadzone: Boolean) {
        super.updateComponents(useLiftDeadzone)
        //lift.updateTeleopNormalPID(false)
    }
}

//fun createAutoBotComponents() =
//    AutoBotComponents(
//        SampleMecanumDrive(hwMap),
//        Camera(),
//    )
//
//data class AutoBotComponents(
//    val drive: SampleMecanumDrive,
//    val camera: Camera,
//) : BaseBotComponents() {
//    override fun updateComponents(useLiftDeadzone: Boolean) {
//        super.updateComponents(useLiftDeadzone)
//        camera.update()
//        drive.update()
//        lift.updateAutoLiftNormalPID()
//    }
//}